server/lib/parties/party-service.ts

```diff
@@ -95,11 +97,26 @@ export default class PartyService {
         )
       }
 
+      if (party.invites.has(invitedUser.id)) {
+        throw new PartyServiceError(
+          PartyServiceErrorCode.InvalidAction,
+          'An invite already exists for this user',
+        )
+      }
+
+      if (party.members.has(invitedUser.id)) {
+        throw new PartyServiceError(
+          PartyServiceErrorCode.InvalidAction,
+          'This user is already a member of this party',
+        )
+      }
+
       party.invites.set(invitedUser.id, invitedUser)
-      this.publishToParty(party.id, {
+      const inviteEventData: PartyInviteEvent = {
         type: 'invite',
         invitedUser,
-      })
+      }
+      this.publishToParty(party.id, inviteEventData)
```

Comment:
Now that you have these things typed, you could change `publishToParty` to:
```ts
private publishToParty<T extends PartyEvent>(partyId: string, data: T)
```

and then you could use it without pulling these into separate fields or needing to declare the event type.